lib/puppet/agent.rb

```diff
@@ -1,6 +1,167 @@
 require 'puppet/application'
 require 'puppet/error'
 
+# A class providing pre/post fork callbacks modeled after those registered
+# by the `pthread_atfork` function.
+# The callbacks defined in this class are no-ops on all platforms except
+# Solaris where they ensure the forked process runs in a different contract
+# than the parent process. This is necessary in order for the child process
+# (an agent run) to be able to restart/stop the puppet agent service without
+# being killed (as a result of the termination of the parent process
+# contract during the service restart/stop).
+# The implementation uses Fiddle to make calls to functions defined in the
+# libcontract library.
+class AtFork
+  require 'fiddle'
+
+  raise LoadError, %q(The loaded fiddle version doesn't define 'Fiddle::Handle') unless defined?(Fiddle::Handle)
```

Comment:
What Fiddle versions don't support this? How likely are we to encounter them in the wild?

If we expect people to run into this error, I'd be interested in making it more graceful (i.e. the AtFork functionality doesn't work but everything else works as before).